### PR TITLE
Support recreating containers

### PR DESCRIPTION
**Pull request by imikushin**
In lieu of #16

### CODE REVIEW

**Comment by imikushin** [comment]
I've addressed the review comments, and there's one "FIXME" left to discuss. When that's addressed, this PR will be good to merge.

**Comment by ibuildthecloud** [comment]
@imikushin We need to do volumes from the old container so that it copies all the same mounts as the older container.  Then we delete the container later.  I believe the behavior of volumes from is that is copies the info from the old container and then you can delete the old container.

**Comment by vdemeester** [comment]
This is how compose works I think. /cc @dnephin

**Comment by dnephin** [comment]
We don't use `volume_from` any more because it ends up copying over some volumes we don't want. The new logic is:

* rename old container to a temporary name
* create a new container with the old volumes (https://github.com/docker/compose/blob/5ed9f9/compose/service.py#L897-L910)
* remove old container with temporary name

So we pick out only the data volumes that are still in the new configuration, and copy those over.

**Comment by ibuildthecloud** [comment]
@dnephin This is close to what we do now in this PR, but we are still using `volumes-from`.  @imikushin Can you do the same as docker-compose?  https://github.com/docker/compose/blob/5ed9f9/compose/service.py#L913-L938  That should be a simple change but we I believe we already to the renaming logic.

**Comment by imikushin** [comment]
I've implemented mounts migration.

**Comment by imikushin** [comment]
Updated volume migration logic and added a test. @ibuildthecloud @dnephin @vdemeester Please take a look.

File: docker/container.go [hunk 1]
```diff
@@ -198,7 +240,7 @@ func (c *Container) Up(imageName string) error {
 
 // OutOfSync checks if the container is out of sync with the service definition.
 // It looks if the the service hash container label is the same as the computed one.
-func (c *Container) OutOfSync() (bool, error) {
+func (c *Container) OutOfSync(imageName string) (bool, error) {
```
Comment (ibuildthecloud) [inline docker/container.go]: Is imageName used?

File: docker/container.go [hunk 2]
```diff
@@ -209,10 +251,28 @@ func (c *Container) OutOfSync() (bool, error) {
...
+	}).Debug("Comparing old and must-be hash")
+	return info.Config.Labels[HASH.Str()] != c.getHash(), nil
+}
+
+func (c *Container) getHash() string {
+	return project.GetServiceHash(c.service.Name(), c.service.Config())
 }
 
-func (c *Container) createContainer(imageName string) (*dockerclient.APIContainers, error) {
+func volumeBinds(volumes map[string]struct{}, container *dockerclient.Container) []string {
```
Comment (ibuildthecloud) [inline docker/container.go]: The python code looks at external: https://github.com/docker/compose/blob/5ed9f9/compose/service.py#L926  Is that covered here?

File: docker/container.go (hunk 1 above)
Comment (imikushin) [inline docker/container.go]: oops! not anymore

File: docker/container.go (hunk 2 above)
Comment (imikushin) [inline docker/container.go]: Yeah, the `volumes` arg here doesn't have external volumes, so no need to check.

**Comment by ibuildthecloud** [comment]
LGTM

**Comment by ibuildthecloud** [comment]
@vdemeester I'm particularly interested in getting this PR merged soon.  We really need this functionality so currently we have a rancher specific fork of libcompose that we ship in Rancher.  We really want to move to upstream but really haven't been able to because of this PR.

There's probably bugs in this or something isn't 100% correct, but I'd rather merge the PR and then iterate on bugs then wait for something that is perfect.

File: docker/container.go [hunk 3]
```diff
@@ -96,6 +97,51 @@ func name(names []string) string {
 	return current[1:]
 }
 
+// Rebuild will delete the container and create a new one with the current configuration
+func (c *Container) Rebuild(imageName string) (*dockerclient.APIContainers, error) {
```
Comment (dnephin) [inline docker/container.go]: I think "build" is the wrong terminology here. We build images, but we create containers. I think it should be `Recreate()`

Comment (dnephin) [inline docker/container.go]: same thing with the error messages below

**Comment by vdemeester** [comment]
I'm ok with, reviewing tomorrow and we'll get this merged soon ;-)

File: docker/container.go [hunk 4]
```diff
@@ -96,6 +97,51 @@ func name(names []string) string {
...
+
+	name := info.Name[1:]
+	newName := fmt.Sprintf("%s_%s", name, info.ID[:12])
+	deleted := false
+	logrus.Debugf("Renaming %s => %s", name, newName)
+	if err := c.client.RenameContainer(dockerclient.RenameContainerOptions{ID: info.ID, Name: newName}); err != nil {
+		logrus.Errorf("Rename failed, deleting %s", name)
+		if err := c.client.RemoveContainer(
+			dockerclient.RemoveContainerOptions{ID: info.ID, Force: true, RemoveVolumes: false}); err != nil {
+			return nil, err
```
Comment (dnephin) [inline docker/container.go]: This is wrong. If rename fails you can't remove the old container. That would cause you to lose your volumes (unless you go in and manually copy them over somewhere, which wasn't possible before 1.9).

File: docker/service.go [hunk 5]
```diff
@@ -186,15 +200,59 @@ func (s *Service) up(imageName string, create bool) error {
...
+		"newRebuildLabel":     newRebuildLabel,
+		"outOfSync":           outOfSync,
+		"ForceRecreate":       s.context.ForceRecreate,
+		"rebuildLabelChanged": rebuildLabelChanged}).Debug("Rebuild values")
+
+	if s.context.ForceRecreate || origRebuildLabel == "always" || rebuildLabelChanged || origRebuildLabel != "false" && outOfSync {
+		logrus.Infof("Rebuilding %s", name)
+		if _, err := c.Rebuild(imageName); err != nil {
+			return err
+		}
```
Comment (dnephin) [inline docker/service.go]: This doesn't look like any logic we have in docker-compose, what's this about?

Comment (dnephin) [inline docker/service.go]: I don't see where this `REBUILD` label is being set on the container.

Why does it matter what the flag was when we originally created the container? I think the only thing that matters is the current flags to `up`.

**Comment by dnephin** [comment]
@ibuildthecloud  I don't think this is ready to be merged. I've left some initial comments

File: docker/container.go (hunk 3 above)
Comment (vdemeester) [inline docker/container.go]: Completely agree :+1:

File: docker/service.go (hunk 5 above)
Comment (ibuildthecloud) [inline docker/service.go]: @dnephin I guess we need to find a different approach then.  So this code was originally taken from RancherOS.  In RancherOS we have containers that must always be recreated on "up".  This is what this logic is about.  By adding a label equal to "always" it basically does force recreate on every up for certain containers.

Comment (ibuildthecloud) [inline docker/service.go]: @dnephin Actually let me explain this logic a bit more.  There are two values you can put for the label REBUILD=always or REBUILD=false.  The context here is specially the "console" container in RancherOS.  The way RancherOS works is that the when the OS boots we essentially do "up".  We have three modes for the console: ephemeral, persistant, "default".

1. ephemeral: An every up we always recreate, the is the equivalent of `--force-recreate`.
1. persistant: Never recreate the container regardless if the compose file changes.  Basically the equivalent of `--no-recreate`
1. default: just normal logic

The logic gets confusing when you recreate container with different values of this REBUILD tag.  For example if REBUILD=false and then you edit the compose file to REBUILD=always or vice versa.  I'll tell you we tested this extensively.

We really need this logic, so if you don't want it in libcompose we will just have to find away of pulling it out.  The key thing here is that we need to essentially set `--force-recreate/--no-recreate` on a specific service in the project and ignore the which mode is set for the current up.

Your call.  I see two paths forward.  We can keep the feature but change the label and terminology to fall more inline with docker-compose, so RECREATE=no or RECREATE=force.  Or we yank it out and just do it in RancherOS.  I think it's a useful feature.

File: docker/container.go (hunk 3 above)
Comment (ibuildthecloud) [inline docker/container.go]: Makes sense.  @imikushin Can you rename everything to be recreate.  I just messed that one up.

File: docker/container.go (hunk 4 above)
Comment (ibuildthecloud) [inline docker/container.go]: Okay, originally the rename was done as "best effort" but I agree that you could loose data.  @imikushin Can you just remove this?  It actually makes the code much cleaner.  Just fail if we can't rename.

**Comment by ibuildthecloud** [comment]
@dnephin Thanks for reviewing.  Just to clarify I don't want to rush this in especially if its wrong.  I just was just saying if maybe a corner case or something wasn't covered we should consider merging.  But you right that there as definitely some wrong things here.

**Comment by imikushin** [comment]
Hopefully, now it's less wrong :)

File: docker/service.go (hunk 5 above)
Comment (dnephin) [inline docker/service.go]: Interesting idea. I agree it does get pretty confusing when you have different containers in a stack that have different recreate policies. I can see a scenario that is just impossible to resolve (at least with links or volumes) where a service is set to RECREATE=no and it has a dependency that is set to RECREATE=always. 

The links will be broken every time, so the RECREATE=no is basically invalid, but a lot of work to detect.

I'd feel a lot better leaving this out of libcompose (at least for now) because it can lead to these kind of impossible scenarios.  With the new parameters I think it should be possible to instrument this from outside of libcompose by setting the label in the config, and inspecting the containers before calling recreate.

File: project/hash.go [hunk 6]
```diff
@@ -11,13 +11,11 @@ import (
...
-// have changed and needs to be rebuild.
-func GetServiceHash(service Service) string {
+// have changed and needs to be recreate.
+func GetServiceHash(name string, config *ServiceConfig) string {
 	hash := sha1.New()
 
-	io.WriteString(hash, fmt.Sprintln(service.Name()))
-
-	config := service.Config()
+	io.WriteString(hash, name)
```
Comment (dnephin) [inline project/hash.go]: I think the config has is missing `ImageId` (it needs to resolve the current id from the name), and we don't track `name` in the hash (any change in name will already be reflected in the labels).

File: cli/command/command.go [hunk 7]
```diff
@@ -221,6 +229,8 @@ func Populate(context *project.Context, c *cli.Context) {
 		context.Log = true
 	} else if c.Command.Name == "up" {
 		context.Log = !c.Bool("d")
+		context.NoRecreate = c.Bool("no-recreate")
+		context.ForceRecreate = c.Bool("force-recreate")
```
Comment (dnephin) [inline cli/command/command.go]: I think these should be [converted to an enum](https://github.com/docker/compose/blob/f3d49e38c0e488194cd507b092111e0bce21e779/compose/cli/main.py#L622) and passed as a parameter to `project.Up()`. They aren't really part of the "project context", they're a parameter for a single operation.

Comment (ibuildthecloud) [inline cli/command/command.go]: While your comment is most likely valid, I think it's out of the scope of this PR.  Basically all options from the CLI are currently set in the Context and not passed as options.  We can make that change, but I don't think it's needed for this PR.

File: project/hash.go (hunk 6 above)
Comment (ibuildthecloud) [inline project/hash.go]: @dnephin Your right, @imikushin something happened here.  In my original PR getHash() took an image name.  You seem to have dropped that, so there is not more logic that checks the imageId has changed.  https://github.com/docker/libcompose/pull/16/files#diff-caac1491e4add415147af39b799d3bdaR258

Comment (ibuildthecloud) [inline project/hash.go]: @imikushin Okay, I figure out what's going on there.  I screwed up.  My original PR had the wrong commit, it wasn't the same logic we are using in RancherOS.  I've cherry picked the right commit and updated #16.  Sorry about that, can you rebase with #16?

**Comment by imikushin** [comment]
Re-rebased @ibuildthecloud 's PR #16 latest incarnation, renamed according to the canon. 
OMG, tests just passed! Please review. 
/cc @vdemeester @dnephin

File: docker/labels.go [hunk 8]
```diff
@@ -11,10 +11,11 @@ type Label string
...
 const (
-	NAME    = Label("io.docker.compose.name")
-	PROJECT = Label("io.docker.compose.project")
-	SERVICE = Label("io.docker.compose.service")
-	HASH    = Label("io.docker.compose.config-hash")
+	NAME     = Label("io.docker.compose.name")
+	PROJECT  = Label("io.docker.compose.project")
+	SERVICE  = Label("io.docker.compose.service")
+	HASH     = Label("io.docker.compose.config-hash")
+	RECREATE = Label("io.docker.compose.recreate")
```
Comment (dnephin) [inline docker/labels.go]: I don't think this label is used anywhere

Comment (imikushin) [inline docker/labels.go]: https://github.com/imikushin/libcompose/blob/rebuild/docker/service.go#L216-L217

Comment (dnephin) [inline docker/labels.go]: I thought we were going to remove that behaviour?

It's not supported by docker-compose, and it's not actually set anywhere from libcompose, which means it relies on something external to set it?

Comment (dnephin) [inline docker/labels.go]: I think a better way to implement it would making the `forceRecreate` param we talk about here: https://github.com/docker/libcompose/pull/89/files#r43012067

Comment (imikushin) [inline docker/labels.go]: How would I mark one particular service as "recreate never" in my compose template? 
That is, when I only need one service to persist and assume others (in the same template) are treated by default.

Comment (ibuildthecloud) [inline docker/labels.go]: @dnephin RECREATE is set by the user.  I don't think you really confirmed that we should remove the functionality, but I think the discussion has been lost by updates to the PR.  We can't just remove the ability for users to control the recreate behavior through labels.  I really like the feature, actually need the feature, but we will just do it in RancherOS code I guess.  

Regarding https://github.com/docker/libcompose/pull/89/files#r43012067 I don't want to change the approach of passing arguments in this PR.  This code is consistent with the rest of the code base so it doesn't make sense to do it differently.  If we think it should be done a different way then lets address that in a different PR.

Comment (ibuildthecloud) [inline docker/labels.go]: @dnephin Sorry there was a terrible typo in that last comment I just fixed.  I should have said "We can just remove the ability"...

Comment (dnephin) [inline docker/labels.go]: >  If we think it should be done a different way then lets address that in a different PR.

I think that's fair, I don't want to overload this PR

Comment (imikushin) [inline docker/labels.go]: @dnephin, Regarding the RECREATE label: true it's not supported by docker-compose now. But if docker-compose adds similar behaviour, it'll need a similar label. 

@ibuildthecloud and I need this feature for RancherOS, because we're using libcompose to manage system services, and some of them (like console) need to be persistent while the others are recreated on boot. Sure we can re-implement parts of libcompose (or the whole thing, as was the case before @ibuildthecloud factored it out from RancherOS and *created* libcompose), but that doesn't sound like progress.

I'd really like to know what's wrong with this feature (marking services with desired recreate behaviour) that makes it unacceptable here. If someone uses this "recreate" label in their compose template with libcompose, does it hurt docker-compose in any way? Sure they might ask docker-compose maintainers to add support for it in a future version, but is it a bad thing?

Another question: is there an official document for libcompose that states it should be a subset of docker-compose, feature-wise?

Comment (dnephin) [inline docker/labels.go]: >  But if docker-compose adds similar behaviour, it'll need a similar label.

It already has similar behaviour. It uses the `--no-recreate` and `--force-recreate` to handle it. With the refactor I've suggested (passing them as arguments to the recreate function, instead of through the context) this would be possible in libcompose as well.

> I'd really like to know what's wrong with this feature (marking services with desired recreate behaviour) that makes it unacceptable here

I think I've alluded to it already, but to re-state:

1. It can lead to impossible situations where you have one service that depends on another, but the service is set to never recreate. When the service is recreated, the service that is set to never recreate breaks because the links or volumes aren't pointing at the correct place anymore. How would you handle that?
2. I think the interface is wrong. It should be handled with parameters. Right now it's a magic label that compose reads but never sets. This is a deviation from all the other labels.
3. It hasn't gone through the standard review process for a new feature. Anything we add to docker-compose has a ticket created and a discussion of the design and implementation. This new feature is being mixed in with a feature parity branch, which is not the right place for it.
4.  It doesn't need to be a compose label, with parameters it could be a rancherOS label, which I think is an argument for keeping it out of compose, at least until there is some more interest in such a feature. 

> Sure we can re-implement parts of libcompose ...

I think with the refactoring I've suggested (to move it to an argument) you won't need to re-implement anything. You'll be able to store the config on the RancherOS side as a label. You can then read the label or config to determine how a conatiner should be recreated.

> is there an official document for libcompose that states it should be a subset of docker-compose, feature-wise?

The README states as the first line "A Go library for Docker Compose. It does everything the command-line tool does". It doesn't say anything about new features really.

I don't think it necessarily has to be a subset of features, but new features should go through the same design review that we have for all projects https://github.com/docker/docker/blob/master/project/REVIEWING.md#workflow

Comment (ibuildthecloud) [inline docker/labels.go]: @dnephin 100% agreed, we will rip out the labels.  I think there is a bit of a confusion in that libcompose already supports `--no-recreate` and `--force-recreate` in this PR.  The problem was that it additionally supported the same behavior from labels.  So we will rip out the labels.  We will ensure that libcompose is a strict subset of docker-compose going forward.  The PR was actually created way back and has been dangling for awhile, so I think it wasn't clear how strictly we would conform to docker-compose at that time.  You've made you point loud and clear, so we will always ensure that we don't add new behavior.  Thanks!

File: docker/container.go (hunk 3 above)
Comment (imikushin) [inline docker/container.go]: Addressed.

File: docker/container.go (hunk 4 above)
Comment (imikushin) [inline docker/container.go]: Addressed.

File: docker/service.go (hunk 5 above)
Comment (imikushin) [inline docker/service.go]: Addressed.

File: project/hash.go (hunk 6 above)
Comment (imikushin) [inline project/hash.go]: Addressed.

File: cli/command/command.go (hunk 7 above)
Comment (imikushin) [inline cli/command/command.go]: Will do in a separate PR.

File: docker/labels.go (hunk 8 above)
Comment (imikushin) [inline docker/labels.go]: Addressed.

**Comment by imikushin** [comment]
@dnephin @ibuildthecloud I believe, I've addressed all your comments. 
Now that the controversial "recreate" label has been removed, is this PR ready to be merged?

**Comment by ibuildthecloud** [comment]
@dnephin I think everything is addressed except your comment to pass the convergence strategy as a parameter.  I understand what you are saying there and agree it's better.  Right now though the internals of how services are traversed and what not is actually not correct I believe and the code is pretty dirty.  I'd like to do a separate PR that makes the internals of the code more in line with convergence strategy in docker-compose.  In short, I think passing the flags as arguments can be done in a separate PR.

**Comment by ibuildthecloud** [comment]
LGTM

File: docker/functions.go [hunk 9]
```diff
@@ -34,3 +34,24 @@ func GetContainerByName(client *dockerclient.Client, name string) (*dockerclient
 
 	return &containers[0], nil
 }
+
+// GetContainerByID looks up the hosts containers with the specified Id and
+// returns it, or an error.
+func GetContainerByID(client *dockerclient.Client, id string) (*dockerclient.APIContainers, error) {
+	containers, err := client.ListContainers(dockerclient.ListContainersOptions{All: true, Size: false})
```
Comment (vdemeester) [inline docker/functions.go]: I wonder why we get all the container and then filter them out ? `ListContainersOptions` and the API has a way to use filter, and one of them is `id`, so we would let `docker` do the hard work here. Something like :

```go
containers, err := client.ListContainers(dockerclient.ListContainersOptions{
    All: true, 
    Filter: map[string][]string{
        "id":id,
    }, 
    Size: false,
})
```

Then, the only thing is to check the length :wink:

File: docker/service.go [hunk 10]
```diff
@@ -46,15 +59,20 @@ func (s *Service) collectContainers() ([]*Container, error) {
 
 	result := []*Container{}
 
+	if len(containers) == 0 {
```
Comment (vdemeester) [inline docker/service.go]: I'm not sure if it's an optimization or not. If `len(containers)` is equal to `0`, the `for` loop won't do a thing anyway right ? :stuck_out_tongue:

File: docker/service.go [hunk 11]
```diff
@@ -186,15 +188,48 @@ func (s *Service) up(imageName string, create bool) error {
...
+		return err
+	}
+	name := containerInfo.Name[1:]
+
+	logrus.WithFields(logrus.Fields{
+		"outOfSync":     outOfSync,
+		"ForceRecreate": s.context.ForceRecreate,
+		"NoRecreate":    s.context.NoRecreate}).Debug("Going to decide if recreate is needed")
+
+	if s.context.ForceRecreate || outOfSync {
```
Comment (vdemeester) [inline docker/service.go]: We could bring that up (before `containerInfo, err := …`) as we already have all information at this point (except the `name` but see comment below) and this would remove one API call (an inspect done in `c.findInfo()`).

File: docker/service.go [hunk 12]
```diff
@@ -186,15 +188,48 @@ func (s *Service) up(imageName string, create bool) error {
...
+		logrus.Infof("Recreating %s", name)
+		if _, err := c.Recreate(imageName); err != nil {
+			return err
+		}
+	} else if outOfSync {
```
Comment (vdemeester) [inline docker/service.go]: Not sure I get this `if` condition : If `outOfSync` is true, no matter `ForceRecreate`, it got into the first one (and did the `Recreate`), and if `ForceRecreate` is true, it got into the first one too.
So I don't see how this `else if …` is ever gonna be executed. (I might miss something :sweat_smile:)

File: integration/up_test.go [hunk 13]
```diff
@@ -25,6 +26,223 @@ func (s *RunSuite) TestUpNotExistService(c *C) {
...
+
+func mountSet(slice []dockerclient.Mount) map[string]bool {
+	result := map[string]bool{}
+	for _, v := range slice {
+		result[fmt.Sprint(v.Source, ":", v.Destination)] = true
+	}
+	return result
+}
+
+func filter(s map[string]bool, f func(x string) bool) map[string]bool {
```
Comment (vdemeester) [inline integration/up_test.go]: This method looks useful enough to get into some `utils` package or something (with a unit test :heart_eyes_cat:.

File: integration/up_test.go [hunk 14]
```diff
@@ -25,6 +26,223 @@ func (s *RunSuite) TestUpNotExistService(c *C) {
...
+	p = s.FromText(c, p, "up", "--no-recreate", `
+	hello:
+	  labels:
+	    key: val
+	  image: busybox
+	  stdin_open: true
+	  tty: true
+	`)
+	cn2 := s.GetContainerByName(c, name)
+	c.Assert(cn.ID, Equals, cn2.ID)
```
Comment (vdemeester) [inline integration/up_test.go]: Shouldn't we also test the labels are not the one in the yaml ?

**Comment by vdemeester** [comment]
Few comments :stuck_out_tongue_closed_eyes:

File: docker/service.go (hunk 12 above)
Comment (imikushin) [inline docker/service.go]: u'r right, it's just a piece of cruft

File: docker/service.go (hunk 10 above)
Comment (imikushin) [inline docker/service.go]: Right, it's useless :scissors:

File: docker/functions.go (hunk 9 above)
Comment (imikushin) [inline docker/functions.go]: How come it's not documented in the Docker API? :space_invader: 
But it's there :smiling_imp:

Comment (vdemeester) [inline docker/functions.go]: Oh was going to say "yeah i is documented" but.. http://docs.docker.com/engine/reference/api/docker_remote_api_v1.21/#list-containers doesn't seem to be up-to-date :sweat_smile:.

File: docker/service.go (hunk 11 above)
Comment (imikushin) [inline docker/service.go]: Thanks! This one nagged me too. Just a tiny bit: apparently not enough. Fixing.

File: integration/up_test.go (hunk 13 above)
Comment (imikushin) [inline integration/up_test.go]: You mean, `filter`? You got it!

**Comment by imikushin** [comment]
@vdemeester Done :wink:

File: integration/up_test.go (hunk 14 above)
Comment (imikushin) [inline integration/up_test.go]: Done!

**Comment by vdemeester** [comment]
LGTM 🐸
Thanks @imikushin

**Comment by imikushin** [comment]
Happy to finally see it merged! 
Cheers :beers:
